### DISCUSSION TITLE
Removing states from a tensor_basis

### DISCUSSION THREAD

**Discussion by PFmix** [QuSpin/QuSpin#612] [Q&A] [answered]
Hello,

I wish to remove some states of a generated tensor_basis to implement a constraint on my Hilbert space, but I do not now how to proceed. Specifically, let's say that I have generated the tensor_basis `basis`:

```
from quspin.operators import hamiltonian
from quspin.basis import spinless_fermion_basis_1d,tensor_basis,boson_basis_1d 
import numpy as np

basis_fermion = spinless_fermion_basis_1d(L=8)
pairbasis = boson_basis_1d(L=20,sps=2,Nb=1)
basis=tensor_basis(basis_fermion,pairbasis)
```

This basis contains a total of 5120 states indexed (amongst other) by their unique array index that I can identify using `print(basis)`.
What I wish to obtain is this same basis (for later use in various function) but without, for example, the states indexed 54, 786, and 2523.

How do I do that, given that `basis` is not an array such that I cannot remove elements from it with `pop()`?

**Comment by mgbukov** [comment]
Hi,

constraints are more easily implemented using the `user_basis`, see [this example](http://quspin.github.io/QuSpin/examples/example14.html#example14-label) and the corresponding tutorial for the [user basis](http://quspin.github.io/QuSpin/user_basis.html#user-basis-label).

**Reply by PFmix** [reply]
Thank you. I was actually hoping not to use `user_basis` as I do not quite understand how it works, especially the example you provide. Assuming I eventually figure it out, do you confirm that it also works for tensor basis that do not have straightforwardly accessible integer representation (to the best of my understanding)?

**Comment by mgbukov** [comment, marked as answer]
`user_basis` will actually replace `tensor_basis` in your code so you won't need the latter anymore; 

btw, when it comes to `user_basis`, just read the tutorial carefully and follow the steps; at the end of the tutorial, there are a few examples of spin chains coded using the `user_basis` so you should be able to grasp how it works without much problem. 

the `user_basis` has indeed a bit of a steep learning curve, but it will very soon pay off to you, because the level of control you gain is manyfold, compared to the other bases classes.

**Reply by PFmix** [reply]
Thank you, I will follow your advice then.
